### ISSUE TITLE
Better IDE Folding (Disable Unfolding)

### ISSUE THREAD

**Issue by digwallace** [YoYoGames/GameMaker-Bugs#4441]
### Is your feature request related to a problem?

It's really killing me when I'm inside a script and I start making a slight adjustment or writing new code, and either literally everything below gets unfolded, or nearly everything.

### Describe the solution you'd like

I really just want unfolding to be disabled. 

### Describe alternatives you've considered

I've considered alternatives and I can't think of any. I'm aware of the fold/unfold all shortcuts, but there are times in my workflow where I want specific things unfolded and may be working on something that touches on multiple functions for example, or I may have portions folded in order to better focus on what's not and finish coding it.

**Comment by digwallace** [comment]
> I am having the same issue and I agree with @digwallace. If I've folded some code, it should remain folded until I explicitly unfold it myself.
> 
> I think this is actually a bug that needs to be fixed rather than a feature that needs to be added. My guess is that the folding feature in the IDE doesn't know how to maintain the folding status of folded regions when it encounters an unfinished line of code.

I agree. I came actually to re-submit this as a bug, because it's kind of insane how bad it's behaving. There's no way this can be intentional behavior.

**Comment by digwallace** [comment]
> I am having the same issue and I agree with @digwallace. If I've folded some code, it should remain folded until I explicitly unfold it myself.
> 
> I think this is actually a bug that needs to be fixed rather than a feature that needs to be added. My guess is that the folding feature in the IDE doesn't know how to maintain the folding status of folded regions when it encounters an unfinished line of code.

I agree. I think you're right about it being a bug because it's actually ridiculous how bad it is.

**Comment by thennothinghappened** [comment]
I'd assume that when the code editor re-parses your code it doesn't know how to maintain contextual information about what your folds map to after the changes - the syntax tree could potentially change quite drastically after your changes, so it needs some smartness added to know what maps to what after that.

**Comment by digwallace** [comment]
> I'd assume that when the code editor re-parses your code it doesn't know how to maintain contextual information about what your folds map to after the changes - the syntax tree could potentially change quite drastically after your changes, so it needs some smartness added to know what maps to what after that.
> 
> Also yeah I think this would be a bug not a feature request.

That could definitely be the case. It's a black box, so I can only guess, but from experience it does feel like it's a feature of the IDE to attempt to expand code where it detects an error, and in the events described, I think it's causing it to detect errors all the way down, resulting in everything getting expanded, but again, just speculation.

Is there an admin or someone who could elevate the label to a bug?

**Comment by digwallace** [comment]
Would it be possible to get a link to the parent ticket it was considered a Duplicate of?, so that we may track its status.

I'm unable to locate it, probably due to being described differently.

**Comment by digwallace** [comment]
Also, I'm almost certain this is due to a memory leak. I keep the IDE open 24/7 usually  and ever since some time in the Fall, it's started to chug and require a restart. When it starts chugging, it's highly likely this will occur, even more so if I do too many actions. I've noticed it also happening when simply renaming a resource and every script open will expand literally everything if there's as much as a single entry of the resource.

**Comment by YYDan** [comment]
Not seeing a matching original feature request either (although there are MANY bug reports that are about extremely similar symptoms/steps).

**Comment by digwallace** [comment]
> you are requesting a new feature that the code-parser be able to read and process all of your code without needing to expand it all in order to see everything

I've no idea how the code for the IDE was planned or written, so I won't presume to speculate on specific details. However I can confirm Godot's editor, Unreal Editor, Notepad++, VS Code, Visual Studio, Unity Editor, Android Studio, and JetBrains' IDEs, all support this feature.

Hope this helps.
